### DISCUSSION TITLE
Kademlia Node discovery after connecting to bootstrap node.

### DISCUSSION THREAD

**Discussion by haider-rs** [libp2p/rust-libp2p#2810] [Q&A] [answered]
Hey all,
I am trying to populate my network from scratch.
I have three nodes Node A, Node A2 and Node B:

I have my Node A as bootstrap node. Now I run Node B which is on different computer different network. It connects with the bootstrap node no problem. If I run another node in bootstrap node machine Node A2 it connects with the bootstrap node also publish gossip msg through bootstrap node but why Node B does not receive Node A2 identify msg is it not the responsibility of Node A to tell node B that Node A2 have joined the network?

thanks

**Comment by haider-rs** [comment, marked as answer]
It takes some time before adding the node in the bucket. make sure you have a static IP and open port. and RUST_LOG=debug to view the procedure better.

**Comment by ghost** [comment]
Hi there, I'm trying to setup a bootstrap node by libp2p rust, but I have found no document to do. I can see it written in Go but in the same module in Rust I don't. Do you have some instructions about building a bootstrap node? I really appreciate that. Thanks in advance.

**Reply by haider-rs** [reply]
have a look at this implementation. 
https://github.com/dariusc93/libp2p-chat-test/blob/master/src/main.rs#L105

you can replace the address of the node with your node address. Any node can be bootstrap node but be sure that your bootnode have a static IP and open port. So other nodes can connect to it

**Reply by ghost** [reply]
Thanks, I'll take a look.

**Reply by thomaseizinger** [reply]
What exactly a bootstrap node is depends a lot on your usecase. In principle, a bootstrap node is just a node with a known, static address and `PeerId` and can thus be embedded in your program's source code / config file.

Which protocols this bootstrap node supports and how you are actually going to bootstrap a network out of it depends on what you want your network to do.

Also see this discussion on the rendezvous protocol and node discovery in general: https://github.com/libp2p/rust-libp2p/discussions/2876

**Reply by ghost** [reply]
> What exactly a bootstrap node is depends a lot on your usecase. In principle, a bootstrap node is just a node with a known, static address and `PeerId` and can thus be embedded in your program's source code / config file.
> 
> Which protocols this bootstrap node supports and how you are actually going to bootstrap a network out of it depends on what you want your network to do.
> 
> Also see this discussion on the rendezvous protocol and node discovery in general: #2876

Yes, but for my personal project I want to build my own bootstrap nodes. I can see in the examples from original library or from the @Haider-Ali-DS 's example gave me they still use ```/dnsaddr/bootstrap.libp2p.io``` with specific bootstrap node ids in ```BOOTNODES```.

In this [Go example](https://medium.com/rahasak/libp2p-pubsub-peer-discovery-with-kademlia-dht-c8b131550ac7) I found that the author can set ```mode``` for his node to ```server``` by calling ```dht.Mode(dht.ModeServer)```. My problem is I don't really want to use those public bootstrap nodes. I was thinking that I can use ```rendezvous``` as you mentioned but at first, I still want to have a bootstrap node myself. I really appreciate if there is some guidance.

Thank you both for the information.

**Reply by thomaseizinger** [reply]
Well, as I mentioned, it depends what your personal project should achieve. Do you want a DHT (distributed hash table)? If yes, then your nodes should probably make use of the `libp2p-kad` module. Is your network partitioned into various nodes with different roles (i.e. providing different services) then rendezvous might be an option where each node can register themselves with the services the provide.

Building a rendezvous point is pretty straight forward. For initial testing, the following example might suffice: https://github.com/libp2p/rust-libp2p/blob/master/protocols/rendezvous/examples/rendezvous_point.rs

**Reply by thomaseizinger** [reply]
Comparisons with the go implementation are sometimes not that helpful because `go-libp2p` uses a completely different architecture. In rust-libp2p, there are no first and second class protocols. Each protocol implements the same interface: `NetworkBehaviour`.

One disadvantage of this modularity is that you actually need to sit down and think, which modules you want to use in your project, because a `Swarm` byitself does not provide any protocol functionality.

**Reply by ghost** [reply]
@thomaseizinger I got it, very helpful to me. I might need to think of my use cases before diving into. I will be back if there is any concerns. Thank you very much.
